### DISCUSSION TITLE
Why does syntax highlighting look different in C/C++ Dimmed Inactive Regions (#ifdef)

### DISCUSSION THREAD

**Discussion by theFireProject** [microsoft/vscode-cpptools#8135] [Q&A] [answered]
I have a C file which contains multiple `#if`/`#elif` preprocessor directives. Visual Studio Code detects the parts of code that are not going to be processed depending on the given conditions. These sections are dimmed in order to indicate that the code is not going to be compiled or executed. That behavior is controlled by the `"C_Cpp.dimInactiveRegions"` setting.

But that's not all. Apart from that, the syntax highlighting is different for the sections of code that are dimmed.

**I want to turn this behavior off so that all of my C code looks the same (not dim, same syntax highlighting), independent from preprocessor directives.** How to achieve that?

You can see that some words are displayed in white inside the dimmed sections of code instead of light blue.

![grafik](https://user-images.githubusercontent.com/61639215/132886682-63d47d6c-d810-4f38-aa9a-4c66f67c7de8.png)

If I disabled the `"C_Cpp.dimInactiveRegions"` setting mentioned above, the syntax highlighting will still differ.

![grafik](https://user-images.githubusercontent.com/61639215/132886101-6885d547-0501-4983-ace9-b975061ae372.png)

Why is that the case?

**Comment by bobbrow** [comment]
The coloring is based on how the file compiles.  Because those sections are inactive, our IntelliSense compiler skips over them and we do not get detailed information about those tokens.  The colors you get in those sections are based solely on the C/C++ grammar and not based on any additional "semantic" information.

**Comment by Colengms** [comment, marked as answer]
Hi @theFireProject .  The differences you are seeing are likely due to [semantic highlighting](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide).  That provides colorization based on semantic analysis (compilation) of the source code.  Blocks of code that are inactive are not compiled, so not considered for semantic highlighting.  You could disable semantic highlighting globally with `editor.semanticHighlighting.enabled`, or just for the C/C++ extension using `"C_Cpp.enhancedColorization": "Disabled"`.

**Reply by theFireProject** [reply]
Thanks for both of your explanations, it makes sense.

But if I disabled the `semanticHightlighting` / `enhancedColorization`, I guess my code would look like this:

![grafik](https://user-images.githubusercontent.com/61639215/132996502-7cca3326-4e76-4471-8740-cf3e62feccc1.png)

Is it somehow possible to make the code look like this: with the more detailed / more colorful syntax highlighting active for all of the code, including the parts that are not compiled? (screenshot is a montage)

![grafik](https://user-images.githubusercontent.com/61639215/132996471-ef643be2-c6a8-4c5b-9536-22abf6e0bd17.png)

**Reply by sean-mcmanus** [reply]
There is no way currently to get the enhancedColorization to be used in the inactive regions, because that code is skipped by the parser code that generates the colorization. It would be theoretically possible to implement via running the IntelliSense processing another time for the inactive region, but we'd probably have to add some new setting for the alternative configuration to use (i.e. configuration settings that make the inactive region active) -- you could file a feature request for that.

**Reply by Colengms** [reply]
Hi @theFireProject .  I think it would be difficult to support semantic colorization in inactive regions because there is not necessarily an expectation that the content of inactive regions can be compiled.  Semantic colorization essentially requires compiling.  A common use of inactive regions is to disable code intended for another platform or build configuration.

Another option, if using C++ source (not C), might be to use `constexpr` conditionals.  That would ensure the conditionals are evaluated at compile time, but would not involve inactive regions.

**Reply by sean-mcmanus** [reply]
@Colengms Yeah, we'd need some new setting like "inactiveConfiguration" -- that would enable inactive enhanced colorization and inactive squiggles (like VS supports for inactive regions for projects with multiple configurations)...although inactive squiggles would require additional VS Code support for some "inactive" type of problem type to match the VS purple squiggles.
